### PR TITLE
Finish Validator client 

### PR DESCRIPTION
**Pull request by GregTheGreek**
Should finalize validator duties.

Targets latest master on `ethereum/eth2.0-specs`

### CODE REVIEW

**Comment by GregTheGreek** [comment]
I'd like to belive this is ready to review for a First itteration on the general structure re-work @wemeetagain

File: validator/validator.ts
Comment (mpetrunic) [inline validator/validator.ts]: use new logger? It's singleton so you can just import it anywhere

Comment (GregTheGreek) [reply validator/validator.ts]: Loollll I forgot about this thanks!

File: validator/block.ts
Comment (mpetrunic) [inline validator/block.ts]:
```javascript
import logger from "../src/logger";


logger.info("log");
```

It's singleton so you can just import and use it, no need to pass it as prop everywhere.

File: validator/validator.ts
Comment (mpetrunic) [inline validator/validator.ts]: that's wierd heh

Comment (GregTheGreek) [reply validator/validator.ts]: find and replace failed me

File: validator/block.ts
Comment (mpetrunic) [reply validator/block.ts]: @GregTheGreek think you missed this comment :)

**Comment by wemeetagain** [comment]
CI tests aren't running for some reason, maybe the merge conflict?

**Comment by GregTheGreek** [comment]
@wemeetagain Should be good to go

File: README.md
Comment (wemeetagain) [inline README.md]: I don't think lodestar-client does anything

File: src/validator/validator.ts
Comment (wemeetagain) [inline src/validator/validator.ts]: we probably want a `stop` method, too

File: src/validator/block.ts
Comment (wemeetagain) [inline src/validator/block.ts:27]: probably should return `block`

File: src/validator/block.ts
Comment (wemeetagain) [inline src/validator/block.ts]: I don't think we will want to build a block right away, just when called upon.
In fact, this probably doesn't need start/stop methods.

File: src/validator/block.ts
Comment (wemeetagain) [inline src/validator/block.ts]: maybe make this public, or make another way to return a block.

File: src/chain/helpers/validatorStatus.ts
Comment (wemeetagain) [inline src/chain/helpers/validatorStatus.ts]: this line seems stale

File: src/validator/validator.ts
Comment (wemeetagain) [inline src/validator/validator.ts:120]: it seems like this is relying on the beacon chain to calculate if the validator is a proposer at a given slot.
I think we should have the validator calculate if they're a proposer for a given slot.

File: src/chain/helpers/validatorStatus.ts
Comment (wemeetagain) [inline src/chain/helpers/validatorStatus.ts:118]: I think this function (and `isProposerAtSlot`) should go under `src/validator` somewhere.
This is something that the validator should calculate for themselves.

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looking good.

I think the main thing is we'll probably want to move most of the validator processing logic into the validator itself, so the beacon node can remain blissfully unaware of validator concerns. It simply serves data its seen, and broadcasts data its given.
I think this means being able to sync the BeaconState from the beacon node, being able to run the state transition logic within the validator.

Maybe these changes can happen in a future PR though.

File: src/validator/block.ts
Comment (GregTheGreek) [reply src/validator/block.ts]: Yeah this makes more sense. Removing start/stop and making this public actually makes a lot more sense.

File: src/validator/block.ts
Comment (GregTheGreek) [reply src/validator/block.ts]: yeah i agree

File: README.md
Comment (GregTheGreek) [reply README.md]: I think you're right!

File: src/validator/validator.ts
Comment (GregTheGreek) [reply src/validator/validator.ts:120]: Right, but that would mean the validator needs to maintain a pruned copy of the state in order to make that calculation. I can check in on how other clients do these opperations, but I'm curious as to what you think.

IMO thats definitely the next iteration anyway - Need to add a DB to the validator to catch slashing conditions and also that would help facilitate this.

File: src/chain/helpers/validatorStatus.ts
Comment (GregTheGreek) [reply src/chain/helpers/validatorStatus.ts:118]: I'm going to hold off on this until we architect a proper storage mechanism for the client

**Review by wemeetagain — DISMISSED** [review]

**Review by mpetrunic — DISMISSED** [review]

**Comment by GregTheGreek** [comment]
@mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme

**Comment by mpetrunic** [comment]
> @mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme

It only counts reviews from people with write access to repo, so I can't help you anyways :)

**Comment by GregTheGreek** [comment]
> > @mpetrunic @wemeetagain reviews went stale do to a merge conflict in the readme
> 
> It only counts reviews from people with write access to repo, so I can't help you anyways :)

Shouldn't be an issue anymore 😄

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**
